compiler/lib/global_flow.ml

```diff
@@ -0,0 +1,624 @@
+                      Domain.others
+                  | Expr (Block _) -> Domain.bot
+                  | Phi _ | Expr _ -> assert false)
+                known
+          | Top ->
+              List.iter
+                ~f:(fun y -> Domain.variable_escape ~update ~st ~approx Escape y)
+                args;
+              Top))
+
+let propagate st ~update approx x =
+  let res = propagate st ~update approx x in
+  match res with
+  | Values { known; _ } when Var.Set.cardinal known >= 200 ->
+      (* When the set of possible values get to large, we give up and
+         just forget about it. This is crucial to make the analysis
+         terminates in a reasonable amount of time. This happens when
+         our analysis is very imprecise (for instance, with
+         [List.map]), so we may not loose too much by doing that. *)
+      if debug () then Format.eprintf "TOP %a@." Var.print x;
+      Domain.approx_escape ~update ~st ~approx Escape res;
+      Top
+  | Values _ ->
+      (match st.variable_may_escape.(Var.idx x) with
+      | (Escape | Escape_constant) as s -> Domain.approx_escape ~update ~st ~approx s res
+      | No -> ());
+      if st.variable_possibly_mutable.(Var.idx x) then Domain.mark_mutable ~update ~st res;
+      res
+  | Top -> Top
+
+module G = Dgraph.Make_Imperative (Var) (Var.ISet) (Var.Tbl)
+module Solver = G.Solver (Domain)
+
+let solver st =
+  let g =
+    { G.domain = st.vars
+    ; G.iter_children = (fun f x -> Var.Set.iter f st.deps.(Var.idx x))
+    }
+  in
+  Solver.f' () g (propagate st)
+
+(****)
+
+type info =
+  { info_defs : def array
+  ; info_approximation : Domain.t Var.Tbl.t
+  ; info_may_escape : bool array
+  }
+
+let f p =
+  let t = Timer.make () in
+  let t1 = Timer.make () in
+  let rets = return_values p in
+  let nv = Var.count () in
+  let vars = Var.ISet.empty () in
+  let deps = Array.make nv Var.Set.empty in
+  let defs = Array.make nv undefined in
+  let variable_may_escape = Array.make nv No in
+  let variable_possibly_mutable = Array.make nv false in
+  let may_escape = Array.make nv No in
+  let possibly_mutable = Array.make nv false in
+  let st =
+    { vars
+    ; deps
+    ; defs
+    ; return_values = rets
+    ; variable_may_escape
+    ; variable_possibly_mutable
+    ; may_escape
+    ; possibly_mutable
+    ; known_cases = Hashtbl.create 16
+    ; applied_functions = Hashtbl.create 16
+    }
+  in
+  program_deps st p;
+  if times ()
+  then Format.eprintf "      global flow analysis (initialize): %a@." Timer.print t1;
+  let t2 = Timer.make () in
+  let approximation = solver st in
+  if times ()
+  then Format.eprintf "      global flow analysis (solve): %a@." Timer.print t2;
+  if times () then Format.eprintf "    global flow analysis: %a@." Timer.print t;
+  if debug ()
+  then
+    Var.ISet.iter
+      (fun x ->
+        let s = Var.Tbl.get approximation x in
+        if true || not (Poly.( = ) s Domain.bot)
```

Comment:
remove the leftover `if true`.
 Also `(Poly.( = ) s Domain.bot)` => `Domain.equal s Domain.bot`